src/core/ddsc/src/dds_write.c

```diff
@@ -335,6 +335,21 @@ dds_return_t dds_writecdr_impl (struct writer *ddsi_wr, struct nn_xpack *xp, str
   // to deliver_locally
   ddsi_serdata_ref (dact);
 
+#ifdef DDS_HAS_SHM
+  if (wr) {
+    if ((wr->m_entity.m_domain->gv.config.enable_shm && iox_pub_has_subscribers(wr->m_iox_pub)) &&
+        (dinp->iox_chunk != NULL))
+    {
+      iceoryx_header_t * ice_hdr = dinp->iox_chunk;
+      ice_hdr->guid = ddsi_wr->e.guid;
+      ice_hdr->tstamp = dinp->timestamp.v;
+      ice_hdr->data_kind = (unsigned char)dinp->kind;
+      ddsi_serdata_get_keyhash (dinp, &ice_hdr->keyhash, false);
+      iox_pub_publish_chunk (wr->m_iox_pub, ice_hdr);
```

Comment:
This, I believe, causes the Iceoryx to take over ownership of the chunk, and so it may be freed from this point onward. That makes it problematic to push the serdata into the WHC and to perform local delivery.

IINM the local readers will get it from Iceoryx instead of via Cyclone's own short-circuiting, so the problem is solvable by looking ahead whether there is a need to insert the data in the WHC and to serialize the data if there is/if there are remote readers.

Iceoryx is only used for volatile data (at least for the time being), and so doing `write_sample_gc` prior to call `iox_pub_publish_chunk` should force serialization if there are readers. So if the corresponding `to_ser`/`to_ser_ref` call caches the serialized representation it should be safe.

Reply:
@eboasson Ah, correct the ownership of the chunk is transferred to Iceoryx after `iox_pub_publish_chunk`, also I don't think the local readers will be updated by Iceoryx (at least I have not been able to see this behavior in the code). I am not sure now how to overcome this problem of updating local readers without any copy!